lib/elixir/lib/kernel.ex

```diff
@@ -3994,6 +3993,97 @@ defmodule Kernel do
+  or outside guards, as per the requirements of `Macro.guard/3`.
+
+  ## Example
+
+      defmodule Integer.Guards do
+        defguard is_even(value) when is_integer(value) and rem(value, 2) == 0
+      end
+
+      defmodule Integer.Utils do
+        import Integer.Guards
+
+        def is_even_guard(value) when is_even(value), do: true
+        def is_even_guard(_), do: false
+
+        def is_even_func(value) do
+          if is_even(value), do: true, else: false
+        end
+
+        def is_large_even(value) when is_even(value) and value > 10, do: true
+        def is_large_even(_), do: false
+      end
+
+  """
+  defmacro defguard(guard, blocks \\ []) do
+    {definition, implementation} = :elixir_utils.extract_guards(guard)
+    case Macro.decompose_call(definition) do
+      {name, args} ->
+        do_defguard name, args, implementation, blocks, __CALLER__
+      _ ->
+        raise ArgumentError, message:
+          "could not extract a definition for defguard from `#{Macro.to_string definition}`,"
+          <> " expected: `defguard name(args...) when implementation...`"
+    end
+  end
+
+  defp do_defguard(call, args, [], [], env) do
+    case :elixir_dispatch.find_import([], call, length(args), env) do
+      false -> # call(args) doesn't exist, probably trying to define a guard without an implementation
+        raise ArgumentError, message:
+          "defguard requires an implementation,"
+            <> " try: `defguard #{Macro.to_string({call, [], args})} when implementation...`"
+      _ -> # call(args) exists, probably trying to use it to define a guard but forgot the signature
+        raise ArgumentError, message:
+          "defguard requires a name,"
+          <> " try: `defguard name(args...) when #{Macro.to_string({call, [], args})}`"
+    end
+  end
+
+  # Provided a multi-line block, suggest the first line as the implementation
+  defp do_defguard(name, args, [], [{_first, {:__block__, _, [first_line | _rest]}}], _env) do
+    raise ArgumentError, message:
+      "defguard does not accept a block,"
+      <> " try: `defguard #{Macro.to_string {name, [], args}} when #{Macro.to_string first_line}`"
+  end
+
+  # Provided a single-line block, probably meant for that to be the implementation
+  defp do_defguard(name, args, [], [{_first, single_line_block}], _env) do
+    raise ArgumentError, message:
+      "defguard does not accept a block,"
+      <> " try: `defguard #{Macro.to_string {name, [], args}} when #{Macro.to_string single_line_block}...`"
+  end
+
+  defp do_defguard(name, args, implementation, [], env) do
+    {args, vars} =  Macro.prewalk(args, [], fn
+      {ref, _meta, context} = var, acc when is_atom(ref) and is_atom(context) ->
+        {var, [var | acc]}
+      node, acc ->
+        {node, acc}
+    end)
+    refs = for {ref, _meta, context} <- vars, do: {ref, context}
+    test_env = %{env | context: :guard, vars: env.vars ++ refs}
```

Comment:
I think `env.vars` here should only be refs. After all, `defguard` starts a new variable scope.